### ISSUE TITLE
CUPS version 2+ prints only 1 copy instead of multiple - postscript

### ISSUE THREAD

**Issue by HonzaK94** [apple/cups#5796]
Content of a postscript file:

%%EndProlog
%%BeginSetup
<< /PageSize [590.625 835.3125] /ImagingBBox null /ManualFeed false **/NumCopies 2** >> setpagedevice.
%%EndSetup
%%Page: 1 1
/pgSave save def
835.3125 0 exch translate 1 -1 scale[72 300 div 0 0 72 300 div 0 0]concat

This is from CUPS **error_log**

d [24/Jun/2020:15:17:29 +0200] cupsdProcessIPPRequest: copies integer '1'

The postscript file works correctly under OpenSUSE 42.1 - 42.3 (CUPS version 1.7.5). On newer versions, CUPS will only print 1 copy instead of a desired count. Or maybe a problem with cups-filters ?

**Comment by jsmeix** [comment]
I assume
https://bugzilla.opensuse.org/show_bug.cgi?id=1173345
is the same issue and the there attached CUPS error_log contains (excerpts):
```
D [25/Jun/2020:11:15:13 +0200] Loading printer Samsung_M2020_Series...
...
I [25/Jun/2020:11:15:45 +0200] [Job 30] Started filter /usr/lib/cups/filter/gstopdf (PID 3204)
I [25/Jun/2020:11:15:45 +0200] [Job 30] Started filter /usr/lib/cups/filter/pdftopdf (PID 3205)
I [25/Jun/2020:11:15:45 +0200] [Job 30] Started filter /usr/lib/cups/filter/gstoraster (PID 3206)
I [25/Jun/2020:11:15:45 +0200] [Job 30] Started filter /usr/lib/cups/filter/rastertospl (PID 3207)
I [25/Jun/2020:11:15:45 +0200] [Job 30] Started backend /usr/lib/cups/backend/usb (PID 3208)
...
D [25/Jun/2020:11:15:45 +0200] [Job 30] Ghostscript command line: gs -dQUIET -dPARANOIDSAFER -dNOPAUSE -dBATCH -dNOINTERPOLATE -dNOMEDIAATTRS -sstdout=%stderr -sOutputFile=%stdout -sDEVICE=pdfwrite -dCompatibilityLevel=1.3 -dAutoRotatePages=/None -dAutoFilterColorImages=false -dNOPLATFONTS -dColorImageFilter=/FlateEncode -dPDFSETTINGS=/printer -dColorConversionStrategy=/LeaveColorUnchanged -r600x600 -dDEVICEWIDTHPOINTS=612 -dDEVICEHEIGHTPOINTS=792 -I/usr/share/cups/fonts -c \'<</.HWMargins[12.500000 12.500000 12.500000 12.500000] /Margins[0 0]>>setpagedevice\' -f -_
D [25/Jun/2020:11:15:45 +0200] [Job 30] Ghostscript command line: gs -dQUIET -dPARANOIDSAFER -dNOPAUSE -dBATCH -dNOINTERPOLATE -dNOMEDIAATTRS -dShowAcroForm -sstdout=%stderr -sOutputFile=%stdout -sDEVICE=cups -r600x600 -dDEVICEWIDTHPOINTS=612 -dDEVICEHEIGHTPOINTS=792 -dcupsBitsPerColor=8 -dcupsColorOrder=0 -dcupsColorSpace=0 -scupsPageSizeName=Letter -I/usr/share/cups/fonts -c \'<</.HWMargins[12.500000 12.500000 12.500000 12.500000] /Margins[0 0]>>setpagedevice\' -f -_
```
and
https://openprinting.org/printer/Samsung/Samsung-m2020
reads that a `Samsung m2020` is not suppotred by free software
and - as far as it seems - the printer driver `rastertospl` belongs
some "Unified Linux Driver" obtained from Samsung, cf.
https://askubuntu.com/questions/512046/how-to-install-rastertospl-in-my-usr-lib-cups-filter

So the issue could be in the printer driver `rastertospl`
or in any of the other filters `gstopdf` `pdftopdf` `gstoraster`
that all belong to the `cups-filters` project
or the issue could be in Ghostscript that is called by those filters.

See
https://en.opensuse.org/SDB:How_to_Report_a_Printing_Issue
what you could do to diagnose the cause of your printing issue,
in particular see the section
"It always helps to simulate printout on a virtual generic PostScript printer".

**Comment by jsmeix** [comment]
In general I wonder if a PostScript ` <</NumCopies 2 >> setpagedevice` instruction
can be expected to work in general on non-PostScript printers.
Perhaps before it had worked more by luck than by intention.
Cf. the section about "Common printing data formats" in
https://en.opensuse.org/Concepts_printing
Perhaps with the PDF workflow PostScript specific things like
PostScript `setpagedevice` instructions may got lost?

**Comment by HonzaK94** [comment]
The problem only appears with specific printers. Some work, some don't on new versions. But everywhere else (means CUPS version >2 and OpenSUSE version >15 every printer works.

**Comment by HonzaK94** [comment]
I am currently trying to simulate printout on a virtual generic PostScript printer but for some reason no file is generated when I input the commands mentioned there.

**Comment by HonzaK94** [comment]
[testq.out.zip](https://github.com/apple/cups/files/4831078/testq.out.zip)
I have managed to simulate printout.

**Comment by HonzaK94** [comment]
%RBINumCopies: 1

**Comment by HonzaK94** [comment]
But why would that problem only occur with newer CUPS and OpenSUSE ?

**Comment by jsmeix** [comment]
`%RBINumCopies: 1` is only a PostSript comment (because of the leading `%`)
and PostSript comments do not affect the final output (because comments
in a programming language do not affect what the program does).

But your
https://github.com/apple/cups/files/4831078/testq.out.zip
also contains (excerpt)
```
<< ... /NumCopies 3 >> setpagedevice
```
so on a PostScript printer your testq.out.zip should result 3 printed sheets
if I understand it correctly how that PostScript instruction is meant to work.

A warm welcome in the hell of print job data processing ;-)

**Comment by HonzaK94** [comment]
Yes that is correct. It should result in 3 printed sheets but only 1 comes out.

**Comment by jsmeix** [comment]
`<< ... /NumCopies 3 >> setpagedevice` should result copies on a PostScript printer
(i.e. when the PostScript program is sent "as is" directly to the PostScript printer)
but I have neither an idea what should happen on non-PostScript printers
nor do I know what actually happens in practice on non-PostScript printers.
I guess what actually happens in practice on non-PostScript printers
depends unpredictably on various obscure conditions.

FYIW:
https://www.ghostscript.com/doc/current/VectorDevices.htm
describes how Ghostscript deals with it, cf. `-dDoNumCopies`

**Comment by HonzaK94** [comment]
https://github.com/OpenPrinting/cups-filters/issues/255 culprit has been found

**Comment by tillkamppeter** [comment]
Fixed OpenPrinting/cups-filters#255. Thank you very much for the bug report. Please test.
